File: lab2/java/SuffixArray.java

import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.ObjectOutputStream;
import java.io.Reader;
import java.io.StringWriter;
import java.io.Writer;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.zip.GZIPInputStream;
import java.util.Arrays;
import java.util.Iterator;
import java.util.NoSuchElementException;
import java.util.Random;

public class SuffixArray {
    String text;
    int[] index;

    Path textFile;
    Path indexFile;

    // Internal constants.
    static final String INDEX_SUFFIX = ".jix";
    static final Charset ENCODING = StandardCharsets.UTF_8;
    static final int GZIP_BUFFERSIZE = 16384;

    public void setText(String text) {
        this.text = text;
        this.textFile = null;
        this.indexFile = null;
        this.index = null;
    }

    public void generateRandomText(int size, String alphabet) {
        Random random = new Random();
        char[] text = new char[size];
        for (int i = 0; i < size; i++) {
            text[i] = alphabet.charAt(random.nextInt(alphabet.length()));
        }
        setText(new String(text));
    }

    public void loadText(String textFile) throws IOException {
        this.textFile = Path.of(textFile);
        this.indexFile = Path.of(textFile + INDEX_SUFFIX);
        if (textFile.endsWith(".gz")) {
            this.text = readGZippedTextfile();
        } else {
            this.text = Files.readString(this.textFile, ENCODING);
        }
        this.index = null;
    }

    private String readGZippedTextfile() throws IOException {
        try (
            InputStream compresesed = Files.newInputStream(this.textFile);
            InputStream uncompressed = new GZIPInputStream(compresesed);
            Reader reader = new InputStreamReader(uncompressed, ENCODING);
            Writer writer = new StringWriter();
        ) {
            char[] buffer = new char[GZIP_BUFFERSIZE];
            int length;
            while ((length = reader.read(buffer)) > 0) {
                writer.write(buffer, 0, length);
            }
            return writer.toString();
        }
    }

    public void loadIndex() throws IOException, ClassNotFoundException {
        try (ObjectInputStream stream = new ObjectInputStream(
                Files.newInputStream(this.indexFile)
            ))
        {
            this.index = (int[]) stream.readObject();
        }
    }

    public int size() {
        return this.text.length();
    }

    public int compareSuffixes(int suffix1, int suffix2) {
        if (suffix1 == suffix2) {
            return 0;
        }
        int end = text.length();
        while (suffix1 < end && suffix2 < end) {
            char ch1 = text.charAt(suffix1);
            char ch2 = text.charAt(suffix2);
            if (ch1 != ch2) {
                return ch1 < ch2 ? -1 : 1;
            }
            suffix1++;
            suffix2++;
        }
        return suffix1 > suffix2 ? -1 : 1;
    }

    public int compareSuffixToValue(int suffix, String value) {
        int end = text.length();
        for (int i = 0; i < value.length(); i++) {
            if (suffix + i >= end) {
                return -1;
            }
            char ch1 = text.charAt(suffix + i);
            char ch2 = value.charAt(i);
            if (ch1 != ch2) {
                return ch1 < ch2 ? -1 : 1;
            }
        }
        return 0;
    }

    public void print(String header) {
        this.print(header, new int[]{0, size()}, "  ");
    }

    public void print(String header, int[] breakpoints, String indicators) {
        this.print(header, breakpoints, indicators, 3, 40);
    }

    public void print(String header, int[] breakpoints, String indicators, int context, int maxSuffix) {
        int digits = Math.max(3, String.valueOf(size()).length());

        System.out.println("--- " + header + " " + "-".repeat(75-header.length()));
        System.out.format("%" + (digits+3) + "s%" + (digits+6) + "s      suffix%n", "index", "textpos");
        String dotdotdot = String.format("%" + (digits+2) + "s%" + (digits+6) + "s", "...", "...");
        int endRange = 0;
        for (int k : breakpoints) {
            int startRange = k - context;
            if (endRange < startRange - 1) {
                System.out.println(dotdotdot);
            } else {
                startRange = endRange;
            }
            endRange = k + context;
            for (int i = startRange; i < endRange; i++) {
                if (0 <= i && i < size()) {
                    char ind = indicators.charAt(0);
                    for (int bp = 0; bp < breakpoints.length; bp++) {
                        if (i >= breakpoints[bp]) 
                            ind = indicators.charAt(bp+1);
                    }
                    int suffixPos = this.index[i];
                    String suffixString = (
                        suffixPos + maxSuffix <= size()
                        ? this.text.substring(suffixPos, suffixPos+maxSuffix) + "..."
                        : this.text.substring(suffixPos)
                    );
                    suffixString = suffixString.replaceAll("\n", "\\\\n").replaceAll("\r", "\\\\n");
                    System.out.format(
                        "%c %" + digits + "d  --> %" + digits + "d  -->  %s%n",
                        ind, i, suffixPos, suffixString
                    );
                }
            }
        }
        if (endRange < size()) {
            System.out.println(dotdotdot);
        }
        System.out.println("-".repeat(80));
        System.out.println();
    }
}

File: lab2/python/suffix_array.py

import gzip
import pickle
import random
from pathlib import Path
from typing import List, Union


class SuffixArray:
    text: str
    index: List[int]

    textFile : Path
    indexFile : Path

    # Internal constants.
    INDEX_SUFFIX = ".pix"
    ENCODING = "utf-8"

    def setText(self, text: str):
        self.text = text
        self.textFile = self.indexFile = self.index = None  # type: ignore

    def generateRandomText(self, size: int, alphabet: str):
        chars = random.choices(alphabet, k=size)
        self.setText("".join(chars))

    def loadText(self, textFile: Union[Path,str]):
        self.textFile = Path(textFile)
        self.indexFile = Path(str(self.textFile) + self.INDEX_SUFFIX)
        openFile = gzip.open if self.textFile.suffix == ".gz" else open
        with openFile(self.textFile, "rt", encoding=self.ENCODING) as IN:  # type: ignore
            self.text = IN.read()
        self.index = None  # type: ignore

    def loadIndex(self):
        with open(self.indexFile, "rb") as IN:
            self.index = pickle.load(IN)

    def size(self) -> int:
        return len(self.text)

    def compareSuffixes(self, suffix1: int, suffix2: int) -> int:
        if suffix1 == suffix2:
            return 0
        text = self.text
        end = len(text)
        while suffix1 < end and suffix2 < end:
            char1 = text[suffix1]
            char2 = text[suffix2]
            if char1 != char2:
                return -1 if char1 < char2 else 1
            suffix1 += 1
            suffix2 += 1
        return -1 if suffix1 > suffix2 else 1


    def print(self, header: str, breakpoints: List[int] = [], 
              indicators: str = "  ", context: int = 3, maxSuffix: int = 40):
        size, text, index = self.size(), self.text, self.index
        if not breakpoints:
            breakpoints = [0, size]
        digits = max(3, len(str(size)))

        print(f"--- {header}", "-" * (75-len(header)))
        print(f"{'index':>{digits+3}s}{'textpos':>{digits+6}s}      suffix")
        dotdotdot = f"{'...':>{digits+2}s}{'...':>{digits+6}s}"

        endRange = 0
        for k in breakpoints:
            startRange = k - context
            if endRange < startRange - 1:
                print(dotdotdot)
            else:
                startRange = endRange
            endRange = k + context
            for i in range(startRange, endRange):
                if 0 <= i < size:
                    ind = indicators[0]
                    for bp in range(len(breakpoints)):
                        if i >= breakpoints[bp]: 
                            ind = indicators[bp+1]
                    suffixPos = index[i]
                    suffixString = (
                        text[suffixPos : suffixPos + maxSuffix] + "..."
                        if suffixPos + maxSuffix <= size else
                        text[suffixPos :]
                    )
                    print(f"{ind} {i:{digits}d}  --> {suffixPos:{digits}d}  -->  {suffixString}")
        if endRange < size:
            print(dotdotdot)
        print("-" * 80)
        print()


